src/customers/Customer.java

```diff
@@ -0,0 +1,682 @@
+
+package customers;
+
+import address.Address;
+import bank.Bank;
+import configs.*;
+import currency.Currency;
+import exceptions.BlockedMailDomainException;
+import exceptions.InvalidPhoneNumberException;
+import exceptions.WeakPasswordException;
+import products.*;
+import services.ExchangeRateService;
+import transaction.TransactionDetail;
+import transaction.TransactionType;
+import utils.Hash;
+
+import java.time.LocalDate;
+import java.time.Period;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public abstract class Customer implements Comparable<Customer>, CustomerOperations {
+    private static int noOfCustomers;
+
+    // nu le fac protected
+    // folosesc getters in subclase cand am nevoie de ele
+    private String hashOfPassword;
+    private String phoneNumber;
+    private String emailAddress;
+    private Address address;
+    private final List<Product> products;
+
+    static {
+        Customer.noOfCustomers = 0;
+    }
+
+    {
+        // each customer has to have at least a current account in RON
+        this.products = new ArrayList<Product>();
+        this.addCurrentAccount("RON");
+
+    }
+
+    @Override
+    public int compareTo(Customer customer) {
+        return customer.products.size() - this.products.size();
+    }
+
+    protected Customer(String password, String phoneNumber, String emailAddress, Address address) {
+        Customer.noOfCustomers += 1;
+
+        try {
+            this.checkPasswordRequirments(password);
+            this.checkMailDomain(emailAddress);
+            this.checkPhoneNumber(phoneNumber);
+        } catch (WeakPasswordException |
+                BlockedMailDomainException |
+                InvalidPhoneNumberException exception) {
+            System.err.println(exception.getMessage());
+            System.exit(Codes.EXIT_ON_ERROR);
+        }
+
+        this.hashOfPassword = Hash.computeHashOfString(password, CustomerConfig.getHashAlgorithm());
+        this.phoneNumber = phoneNumber;
+        this.emailAddress = emailAddress;
+        this.address = address;
+
+    }
+
+    public abstract String getCustomerName();
+
+    public abstract String getCustomerUniqueID();
+
+    public abstract LocalDate getBirthDay();
+
+    public abstract String getUniqueID();
+
+    public int getAge() {
+        LocalDate currentDate = LocalDate.now();
+        LocalDate birthDate = this.getBirthDay();
+
+        return Period.between(birthDate, currentDate).getYears();
+    }
+
+    @Override
+    public void addCurrentAccount(String currencyCode) {
```

Comment:
you should try to  avoid having business logic in an entity class - is good that you defined your abstract methods in the interface CustomerOperations, but what you can refactor: create a CustomerOperationsService which implements the methods from the interface and when you will need to call the methods from Main you will instantiate the  CustomerOperationsService and call the methods from it. This way you decouple your entity class which should contain in general only the fields, constructors, setters, getters, hashCode, equals, toString - think of the entity class as a DTO - data transfer object and think of the Service as the place where you define the logic/methods specific for that DTO

Reply:
O sa tin cont de aceste lucruri de acum inainte. Nu mai am timp sa modific in cod.